### PR TITLE
Fold __is__ and unchecked_cast of derefine

### PR DESCRIPTION
**Pull request by ljfitz**
The added e2e maxpool testcase from #545 was not getting a static shape
due to an unfolded prim.If when RefineTypes was called. This was because
of unfolded torch.iaten.__is__ and torch.prim.unchecked_cast operators
with torch.derefine operands.

### CODE REVIEW

**Review by silvasean — APPROVED** [review]
Thanks :)

**Comment by ljfitz** [comment]
I'm having trouble reproducing the `Build and Test` failure locally. Any tips as to what I might not have updated locally to match the testing setup? I ran `pip install -r requirements.txt --upgrade`, and `git status` doesn't show any local modifications.

**Comment by cathyzhyi** [comment]
> I'm having trouble reproducing the `Build and Test` failure locally. Any tips as to what I might not have updated locally to match the testing setup? I ran `pip install -r requirements.txt --upgrade`, and `git status` doesn't show any local modifications.

Seems the latest pytorch changed the interface for prim::RaiseExeception to `prim::RaiseException : (str, str?) -> ()'`. After updating the pytorch, running `./build_tools/update_torch_ods.sh` shows the following
```
Traceback (most recent call last):
  File "/usr/local/google/home/cathyzhyi/anaconda3/envs/torch-mlir-run/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/google/home/cathyzhyi/anaconda3/envs/torch-mlir-run/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/usr/local/google/home/cathyzhyi/npcomp/mlir-npcomp/build/tools/torch-mlir/python_packages/torch_mlir/torch_mlir/dialects/torch/importer/jit_ir/build_tools/torch_ods_gen.py", line 724, in <module>
    main(args)
  File "/usr/local/google/home/cathyzhyi/npcomp/mlir-npcomp/build/tools/torch-mlir/python_packages/torch_mlir/torch_mlir/dialects/torch/importer/jit_ir/build_tools/torch_ods_gen.py", line 703, in main
    emit_prim_ops(args.torch_ir_dir, registry)
  File "/usr/local/google/home/cathyzhyi/npcomp/mlir-npcomp/build/tools/torch-mlir/python_packages/torch_mlir/torch_mlir/dialects/torch/importer/jit_ir/build_tools/torch_ods_gen.py", line 418, in emit_prim_ops
    emit("prim::RaiseException : (str) -> ()")
  File "/usr/local/google/home/cathyzhyi/npcomp/mlir-npcomp/build/tools/torch-mlir/python_packages/torch_mlir/torch_mlir/dialects/torch/importer/jit_ir/build_tools/torch_ods_gen.py", line 406, in emit
    emit_op(registry[key], f, **kwargs)
  File "/usr/local/google/home/cathyzhyi/npcomp/mlir-npcomp/build/tools/torch-mlir/python_packages/torch_mlir/torch_mlir/dialects/torch/importer/jit_ir/build_tools/torch_ods_gen.py", line 198, in __getitem__
    return self.by_unique_key[key]
KeyError: 'prim::RaiseException : (str) -> ()'
```  

Can you help to fix this line here https://github.com/llvm/mlir-npcomp/blob/52ed3313b49b8f56a6dac270ee334c6dffebcaf2/python/torch_mlir/dialects/torch/importer/jit_ir/build_tools/torch_ods_gen.py#L416

**Comment by ljfitz** [comment]
> After updating the pytorch, running `./build_tools/update_torch_ods.sh` shows the following

It doesn't for me. I wonder if PyTorch has stopped supporting python 3.6. I'll try building with something newer.

**Comment by cathyzhyi** [comment]
I realize it's a bit more than just fix that one line. Fixed in https://github.com/llvm/torch-mlir/pull/549

**Comment by ljfitz** [comment]
> I realize it's a bit more than just fix that one line. Fixed in #549

I just saw your message after I worked out that PyTorch has not been updated for python 3.6 since 20211231 (not surprising given https://endoflife.date/python), managed to reproduce and work out how to make the change.

@cathyzhyi - I'll unwind my `prim::RaiseException` change and wait for #549 to be merged.

**Comment by sjarus** [comment]
Thank you for these fixes @ljfitz and @cathyzhyi! I'm looking forward to trying this out with my maxpool/adaptive avgpool TOSA implementations once it's merged.

**Pull request merged**
